### PR TITLE
Add Hooks support to podman

### PR DESCRIPTION
**Pull request by rhatdan**
Also mv libkpod config into libpod.  Still needs more cleanup,
but this gets us to one library.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

File: libpod/config.go [hunk 1]
```diff
@@ -81,13 +81,13 @@ type RootConfig struct {
...
-	// File-based locking is required when multiple users of libkpod are
+	// File-based locking is required when multiple users of libpod are
 	// present on the same system
 	FileLocking bool `toml:"file_locking"`
 }
 
-// RuntimeConfig represents the "crio.runtime" TOML config table.
-type RuntimeConfig struct {
+// CRIOConfig represents the "crio.runtime" TOML config table.
+type CRIOConfig struct {
```
Comment (mheon) [inline libpod/config.go]: I would be in favor of just dropping this config entirely. We only really use it in getRuntime, and it shouldn't be hard to parse the options ourselves. We might want a dedicated libpod config file similar to c/storage and c/image, but I don't think this is it.

File: libpod/config.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package libkpod
+package libpod
```
Comment (TomSweeneyRedHat) [inline libpod/config.go]: Are we changing this to libpodman eventually?

Comment (mheon) [reply libpod/config.go]: I think we'll be keeping the libpod name

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM, but libostree is causing havoc with the tests.

File: libpod/config.go (hunk 2 above)
Comment (rhatdan) [reply libpod/config.go]: Yes podman is the tool, libpod is the library.

File: libpod/config.go (hunk 1 above)
Comment (rhatdan) [reply libpod/config.go]: I think we should do this in a different PR.  We will need to supply or move the handling of crio.conf back into crio, once we remove libkpod.

**Comment by rhatdan** [comment]
@mheon Tests are passing,

File: libpod/config.go (hunk 1 above)
Comment (mheon) [reply libpod/config.go]: Fair enough. Can you throw a comment at the top of the file saying these configs are deprecated and going away at some point?

File: libpod/container.go [hunk 3]
```diff
@@ -534,6 +538,89 @@ func (c *Container) refresh() error {
...
 
+// GetCommand returns command to be executes inside of container
+// ToDO need to add Command handling to libpod
+func (c *Container) GetCommand() string {
+	return ""
+}
+
+// GetAnnotations returns annotations associated with container
+// ToDO need to add Annotation handling to libpod
+func (c *Container) GetAnnotations() []string {
```
Comment (mheon) [inline libpod/container.go]: We already have this. They're in the OCI spec, which you can get with container.Spec()

File: libpod/container.go [hunk 4]
```diff
@@ -534,6 +538,89 @@ func (c *Container) refresh() error {
 	return nil
 }
 
+// GetCommand returns command to be executes inside of container
+// ToDO need to add Command handling to libpod
+func (c *Container) GetCommand() string {
```
Comment (mheon) [inline libpod/container.go]: Should also be able to get this from the OCI spec, which we do store.

Comment (mheon) [reply libpod/container.go]: I think I'd prefer if we get rid of this and GetAnnotations() and just have people calling libpod get the spec and get the things they need out of it

File: libpod/container.go [hunk 5]
```diff
@@ -151,6 +153,8 @@ type ContainerConfig struct {
 	SharedNamespaceMap map[string]string `json:"sharedNamespaces"`
 	// Time container was created
 	CreatedTime time.Time `json:"createdTime"`
+	// HooksDirPath location of oci hooks config files
+	HooksDirPath string `json:"hooks_dir_path"`
```
Comment (mheon) [inline libpod/container.go]: Is this wired into the database? I think it needs to be, so we can support `podman create` -> `podman start` - start will need to know where the hooks live

Comment (mheon) [reply libpod/container.go]: Also, is this something we should set per-container? I feel like we should be doing this for the entire runtime, not per-container.

File: libpod/container.go [hunk 6]
```diff
@@ -534,6 +538,89 @@ func (c *Container) refresh() error {
...
+
+// GetAnnotations returns annotations associated with container
+// ToDO need to add Annotation handling to libpod
+func (c *Container) GetAnnotations() []string {
+	var annotations []string
+	return annotations
+}
+
+func (c *Container) setupOCIHooks(g *generate.Generator) error {
+	command := c.GetCommand()
```
Comment (mheon) [inline libpod/container.go]: c.config.Spec.Process.Args (99% sure)

File: libpod/config.go (hunk 2 above)
Comment (TomSweeneyRedHat) [reply libpod/config.go]: kk, I was just wondering if we should move to libpodman.  I'm good with the shorter libpod, but didn't know if that would cause confusion or name collision.

File: libpod/container.go [hunk 7]
```diff
@@ -534,6 +538,89 @@ func (c *Container) refresh() error {
...
+	annotations := c.GetAnnotations()
```
Comment (mheon) [inline libpod/container.go]: c.config.Spec.Annotations

**Comment by rhatdan** [comment]
I am now blocked on https://github.com/kubernetes-incubator/cri-o/pull/1244

File: libpod/container.go (hunk 4 above)
Comment (rhatdan) [reply libpod/container.go]: Fixed

File: libpod/container.go (hunk 3 above)
Comment (rhatdan) [reply libpod/container.go]: Fixed

File: libpod/container.go (hunk 5 above)
Comment (rhatdan) [reply libpod/container.go]: It can be specified per container, because of the way podman works. But ordinarily it should be for the entire runtime.

File: libpod/container.go (hunk 6 above)
Comment (rhatdan) [reply libpod/container.go]: I am now checking against Args[0]

File: libpod/container.go (hunk 7 above)
Comment (rhatdan) [reply libpod/container.go]: Done

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
@mheon @TomSweeneyRedHat @baude PTAL

File: libpod/container_api.go [hunk 8]
```diff
@@ -745,3 +747,77 @@ func (c *Container) Sync() error {
...
+		case "prestart":
+			g.AddPreStartHook(h)
+
+		case "poststart":
+			g.AddPostStartHook(h)
+
+		case "poststop":
+			g.AddPostStopHook(h)
+		}
+	}
```
Comment (TomSweeneyRedHat) [inline libpod/container_api.go]: Should we have  a default with at least a debug statement?

File: libpod/container.go [hunk 9]
```diff
@@ -236,8 +236,12 @@ type ContainerConfig struct {
 	CreatedTime time.Time `json:"createdTime"`
 	// Cgroup parent of the container
 	CgroupParent string `json:"cgroupParent"`
+	// HooksDirPath location of oci hooks config files
+	HooksDirPath string `json:"hooks_dir_path"`
 
 	// TODO log options - logpath for plaintext, others for log drivers
+	// TODO save log location here and pass into OCI code
```
Comment (mheon) [inline libpod/container.go]: These comments should be removed

File: libpod/container.go [hunk 10]
```diff
@@ -236,8 +236,12 @@ type ContainerConfig struct {
 	CreatedTime time.Time `json:"createdTime"`
 	// Cgroup parent of the container
 	CgroupParent string `json:"cgroupParent"`
+	// HooksDirPath location of oci hooks config files
+	HooksDirPath string `json:"hooks_dir_path"`
```
Comment (mheon) [inline libpod/container.go]: This needs DB changes to persist across reboot

Comment (mheon) [reply libpod/container.go]: Also, should we set a default in NewContainer()?

File: libpod/config.go [hunk 11]
```diff
@@ -1,4 +1,4 @@
...
 
 import (
```
Comment (mheon) [inline libpod/config.go]: This file really needs a header comment of some kind saying it is deprecated and going away as soon as we can find the time to remove it

File: libpod/container_api.go (hunk 8 above)
Comment (mheon) [reply libpod/container_api.go]: Agree, we should handle default in some way. Probably worth erroring, because the hook has not been added to any stage?

Comment (rhatdan) [reply libpod/container_api.go]: Good point fixed.

File: pkg/hooks/hooks.go [hunk 12]
```diff
@@ -98,4 +97,27 @@ func readHooks(hooksPath string, hooks map[string]HookParams) error {
 	}
 	return nil
 }
-*/
+
+var hooksMap map[string]HookParams
```
Comment (mheon) [inline pkg/hooks/hooks.go]: Can we store this in the runtime instead of as a global?

File: libpod/container.go (hunk 9 above)
Comment (rhatdan) [reply libpod/container.go]: Yes I agree
fixed.

File: libpod/container_api.go [hunk 13]
```diff
@@ -745,3 +747,77 @@ func (c *Container) Sync() error {
 
 	return nil
 }
+
+func (c *Container) setupOCIHooks(g *generate.Generator) error {
```
Comment (mheon) [inline libpod/container_api.go]: Is this ever called when we create the container? I don't think it is

Comment (mheon) [reply libpod/container_api.go]: Also, this is a private function, stick is in container_internal.go

File: libpod/container_api.go [hunk 14]
```diff
@@ -745,3 +747,77 @@ func (c *Container) Sync() error {
...
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// addOCIHook look for hooks programs installed in hooksDirPath and add them to spec
+func (c *Container) addOCIHook(g *generate.Generator, hook hooks.HookParams) error {
```
Comment (mheon) [inline libpod/container_api.go]: Private function, stick in container_internal.go

**Comment by mheon** [comment]
@rhatdan I think you accidentally rebased out the parts where hooks are called

File: libpod/container.go (hunk 10 above)
Comment (rhatdan) [reply libpod/container.go]: Ok moved it around. this directory is just for test purposes, Undocumented, so I guess we can store it in the database, but maybe it should just be ignored in DB,

File: pkg/hooks/hooks.go (hunk 12 above)
Comment (rhatdan) [reply pkg/hooks/hooks.go]: Much better idea.

File: libpod/container_api.go (hunk 13 above)
Comment (rhatdan) [reply libpod/container_api.go]: Fixed.

File: libpod/container_api.go (hunk 14 above)
Comment (rhatdan) [reply libpod/container_api.go]: Fixed

**Comment by rhatdan** [comment]
@mheon PTAL

**Comment by mheon** [comment]
@rhatdan It still doesn't look like SetupHooks() is being called anywhere. Otherwise, it looks fine. If this is intentional I'm good to merge.

**Comment by rhatdan** [comment]
Well sadly this is still not fully working.  Seems like runc is not actually executing the hooks.  Will continue looking into this.  Now runs a test to make sure the hooks fires.

**Comment by mheon** [comment]
bot, retest this please

**Comment by baude** [comment]
LGTM

File: libpod/container_internal.go [hunk 15]
```diff
@@ -1103,3 +1108,58 @@ func (c *Container) saveSpec(spec *spec.Spec) error {
...
+				continue
+			}
+			if match {
+				if err := addHook(hook); err != nil {
+					return err
+				}
+			}
+		}
+		annotations := c.Spec().Annotations
+		for _, annotationRegex := range hook.Annotations {
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal.go]: It would be nice if one of these were a map.  If we expect either one to get rather large at some point, it might be worthwhile to create a map object and then loop through once.  But not a change for now, just something to mull over.

File: libpod/container_internal.go [hunk 16]
```diff
@@ -1103,3 +1108,58 @@ func (c *Container) saveSpec(spec *spec.Spec) error {
...
+			if err := hooks.AddOCIHook(g, hook); err != nil {
+				return err
+			}
+			addedHooks[hook.Hook] = struct{}{}
+		}
+		return nil
+	}
+	for _, hook := range ocihooks {
+		logrus.Debugf("SetupOCIHooks", hook)
+		if hook.HasBindMounts && len(c.config.Mounts) > 0 {
```
Comment (mheon) [inline libpod/container_internal.go]: c.config.Mounts isn't out bind mounts - c.config.Spec.Mounts would be

File: pkg/hooks/hooks.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+	if hooksPath == DefaultHooksDir {
+		if err := readHooks(OverrideHooksDir, hooksMap); err != nil {
+			return nil, err
+		}
+	}
+
+	return hooksMap, nil
+}
+
+// AddOCIHook uses generate to create an OCI specification sing the included hook
```
Comment (TomSweeneyRedHat) [inline pkg/hooks/hooks.go]: nit 'sing the included' to 'using the included'

File: libpod/container_internal.go (hunk 16 above)
Comment (rhatdan) [reply libpod/container_internal.go]: Fixed, Thanks.

**Comment by mheon** [comment]
LGTM

**Comment by rhatdan** [comment]
@rh-atomic-bot r=mheon

File: libpod/container_internal.go [hunk 18]
```diff
@@ -1103,3 +1108,58 @@ func (c *Container) saveSpec(spec *spec.Spec) error {
...
+			if err := hooks.AddOCIHook(g, hook); err != nil {
+				return err
+			}
+			addedHooks[hook.Hook] = struct{}{}
+		}
+		return nil
+	}
+	for _, hook := range ocihooks {
+		logrus.Debugf("SetupOCIHooks", hook)
+		if hook.HasBindMounts && len(c.config.Spec.Mounts) > 0 {
```
Comment (wking) [inline libpod/container_internal.go]: I'm looking at porting kubernetes-incubator/cri-o#1517 to libpod now that kubernetes-incubator/cri-o#1537 has demonstrated vendoring libpod code there, and this line doesn't look right to me.  Shouldn't we be walking `Mounts` looking for `bind` or `rbind` in `Options`?  As it stands this is more like `hasMountsAtAll` instead of `hasBindMounts`.  Can I fix that in my hook reroll, or do we need to preserve the has-mounts-at-all logic here for some reason?

Comment (rhatdan) [reply libpod/container_internal.go]: No I think we should be looking at bind mounts, this is a bug.

Comment (mheon) [reply libpod/container_internal.go]: @wking `c.state.BindMounts` represents mounts that we add to the config before writing it to disk. They are not present in `c.config.Spec.Mounts` which holds the spec the user passed in.

`libpod` accepts an OCI spec when creating a container and uses it mostly unmodified, but we add a few environment variables, annotations, mounts (the aforementioned `/dev/shm` among others), as well as setting the location of the rootfs (not known before we create the container and mount the image)

Comment (mheon) [reply libpod/container_internal.go]: @wking `libpod` doesn't know about `--volume` - it can only distinguish between mounts in the passed-in OCI spec, and mounts it adds itself. Mounts added within `libpod` are stored in `c.state.BindMounts` while the OCI spec, `c.config.Spec`, holds the mounts passed in with the spec. The spec itself is created by parsing the `podman create` or `podman run` command line, and I don't believe it adds any bind mounts not requested by the user via `--volume` (though it's been awhile since I looked through that code).

Comment (mheon) [reply libpod/container_internal.go]: For image volumes - those presently are added separately from both `c.state.BindMounts` and the OCI spec, though I want to refactor them to be stored in `c.state.BindMounts` as well. They are added at the same time as `c.state.BindMounts` is added to the spec, but they are put directly into the final spec we write to disk, skipping `c.state.BindMounts`

Comment (mheon) [reply libpod/container_internal.go]: That perhaps came out a bit more confusing than intended, so let me clarify...

We accept an OCI spec when we create a container in `libpod`. This is not the final spec we use to create the container, as `libpod` adds a few things to it before writing it to disk (`generateSpec()` produces the final spec we write). We add a few mounts during this stage - those included in `c.state.BindMounts` and image volume mounts (if requested). These are never present in `c.config.Spec` which is immutable after being passed in by the user, but will be in the spec we write to disk.

Comment (rhatdan) [reply libpod/container_internal.go]: Well the object is only to cover --volume mounts (bind) mounts. oci-umount is looking at anything handed into the container runtime that is mounted in addition to the mount points that the runtime creates.

Comment (wking) [reply libpod/container_internal.go]: Fix in flight with #700.
